Adds cache usage information when streaming from Anthropic

Penagwin opened the pull request:
## Title

Adds cache usage information when streaming from Anthropic
## Relevant issues

Fixes #5364

## Type

🐛 Bug Fix


## Changes
I saw people were discussing what format to add the cache usage information in https://github.com/BerriAI/litellm/issues/5285

Based on the current support for the Anthropic cache usage information when not using streaming, and also the comment about wanting to make the cache support for Google's Vertex 
https://github.com/BerriAI/litellm/issues/5213#issuecomment-2290537202 I kept the anthropic format

It currently only shows if the llm provider is `anthropic`, but should be easy to add for others. It also only shows if `cache_creation_input_tokens` and `cache_read_input_tokens` were present in any chunks, e.g. only if you set `anthropic-beta: prompt-caching-2024-07-31`

The main thing I don't like about my implementation is that the current typing I think uses `ChatCompletionUsageBlock` from the OpenAI type declaration.

https://github.com/BerriAI/litellm/blob/78a2013e5136e87b443522c17be8111568802fc7/litellm/types/llms/openai.py#L453


## Testing
Are the current pytests broken right now?

I can confirm that anthropic with streaming (with and without the `anthropic-beta` header), without streaming, and gpt-4o with and without streaming works.

Commit:
Adds cache usage information when streaming from Anthropic

Commit:
Fixes the cache usage incorrectly showing for non-streamed openai calls

Commit:
Merge remote-tracking branch 'origin/main' into anthropic_stream_cache_results

krrishdholakia commented on litellm/types/llms/openai.py:
```diff
@@ -454,7 +454,8 @@ class ChatCompletionUsageBlock(TypedDict):
     completion_tokens: int
     total_tokens: int
-
+    cache_creation_input_tokens: Optional[int]
```
it might be better to inherit from this class in types/utils.py, instead of modifying the openai block @Penagwin

Penagwin replied:
I removed the changes to the openai types

krrishdholakia commented on litellm/types/utils.py:
```diff
@@ -457,14 +457,19 @@ def __init__(
             "prompt_tokens": prompt_tokens or 0,
             "completion_tokens": completion_tokens or 0,
             "total_tokens": total_tokens or 0,
         }
```
this usage block is what i'd expect to be updated with having: 
>  cache_creation_input_tokens: Optional[int] = None,
        cache_read_input_tokens: Optional[int] = None,

since they're litellm-specific, not openai-specific params

krrishdholakia commented on litellm/utils.py:
```diff
@@ -9724,6 +9724,8 @@ def chunk_creator(self, chunk):
                             "completion_tokens"
                         ],
                         total_tokens=anthropic_response_obj["usage"]["total_tokens"],
+                        cache_read_input_tokens=anthropic_response_obj['usage']['cache_read_input_tokens'],
```
this is not a guarantee - this block is used for any streaming iterator which returns a GenericStreamingChunk 

You'd also want to update the GenericStreamingChunk type, if you expect this to be a required field

Penagwin replied:
> this is not a guarantee - this block is used for any streaming iterator which returns a GenericStreamingChunk
> 
> You'd also want to update the GenericStreamingChunk type, if you expect this to be a required field

@krrishdholakia Would Mistral be one that uses `GenericStreamingChunk`?

krrishdholakia replied:
no - it's used by vertex ai, gemini (google ai studio), bedrock, and anthropic

Penagwin replied:
I switched to `.get()`

krrishdholakia commented:
Can you add a test for this, to `test_completion.py` @Penagwin

Commit:
Merge remote-tracking branch 'origin/main' into anthropic_stream_cache_results

Commit:
Merge remote-tracking branch 'origin/main' into anthropic_stream_cache_results

Commit:
Removed the changes from the OpenAI types

Commit:
Check for the existence of the cache key's because bedrock/vertex use it

Commit:
Adds a basic test to check that the cache usage numbers are present

Penagwin commented:
@krrishdholakia I think I addressed most of your comments. The test I added works for anthropic.

I don't have bedrock/vertex keys handy, could you double check to see if this breaks them?

thderoo commented:
Any updates on this ? I would really like to have this feature. Thank you

krrishdholakia commented:
@thderoo This is already live - https://github.com/BerriAI/litellm/commit/f9e6507cd1524a7c287abfe75f82594ed1723244


If you don't see it, can you please bump to latest, and if the issue persists - share steps to repro the issue?
